### ISSUE TITLE
Backwards order in current area and unnecessary/unpractical backtracking.

### ISSUE THREAD

**Issue by elreymu** [theonlydude/RandomMetroidSolver#6]
Hello i don't know if this is new but i have noticed that the solver often skips major and minor items that are completely accessible and then backtracks to them much later on in the game.
Also it seems like some items in the same area are being picked up in reverse order which will make things confusing, i will explain further with this seed as an example:
(Tournament seed 6640613 on very hard difficulty (same happens in other difficulties aswell)

In this solution, the solver goes for the charge beam location right after the "Missile (blue Brinstar bottom)" , from there the next major item location is bomb but it skips it completely (alcatraz escape technique is turned on but still it skips it) then after brinstar and norfair it comes back to crateria for that bomb which could have been avoided by just picking it up on the way to charge beam.
Then after charge beam location, it tells you to go to the etecoons super missile (same as bomb, unnecessary backtracking since you could pick it up on your way to charge beam), then to mockball reserve (again, backtracking as if the order was upside down)
backtracking like that happens a lot with a lot of items even at the same area.
another example from that seed is that after bomb location it suggests you to go to the terminator room and then backtrack to the gauntlet e-tank location (it would simpler to go first to the gauntlet and then to the terminator room considering you're near the bomb location)
Also on the why it can get things confusing here's this another example:
Later on in this same seed tells you to go to the speedbooster location, but it isn't possible with your current items (as far as i know), same with the wave beam location, turns out you have to scroll down where it says you have to go to crocomire to get both speedbooster and grapple which now makes it possible to reach speedbooster and wave locations.
So basically, you're supposed to go to crocomire first before you can go to speedbooster and wave locations but the solver displays that order in reverse (and the problem is that order is not possible to do) which can confuse most people unless they have more knowledge about the game.
this issue/behavior repeats with all of the seeds i have tried.

I hope this helps to further improve the solver.
Please also add "SpringBall-Jump from wall" as a way to access "Reserve tank, Norfair"
i think that method is not there so would be nice to have the solver use it too.

**Comment by flohgh** [comment]
Hi, thank you for the suggestions, we'll add spring ball jump from wall as a technique to access Norfair Reserve.

Regardng the spoiler log, you should not take it as a path to follow, but more like a bonus information on the path taken by an algorithm that doesn't actually go through the ROM. What I mean by that is every item is taken as if you were at the landing site. This is mainly because our code is largely based on the actual randomizer code, which proceeds like this to place items. Our tool is only meant to tell you if a ROM is actually finishable, and approximately how hard it will be for you. The spoiler log is more of a bonus.

That being said the algorithm does its best to be "realistic" and stay in the same area. But the path taken within the given area does not take into account the actual distance/backtracking between items. That's why it does not go to crateria items at the beggining, because blue Brinstar at the beggining of the game is indeed Brinstar, not Crateria. Though when enough items are collected, it goes in a boss rush, which is pretty fun to see.

The fact that every item's difficulty is the same as if it was picked at the landing site is the main problem of the current algorithm. For example, Ridley's difficulty will be Lower Norfair access difficulty+Worst Room+Ridley fight instead of just the actual Ridley fight. We made some adjustments for this not to have too much of an impact on the actual difficulty score. 1) The ROM's difficulty is the difficulty of the hardest item 2) our difficulty categories scores are chosen carefully for the problem not to show.

As we might have mentioned before we are currently working on a real navigation algorithm that actually "plays" the ROM and "move" through it, but it is not ready yet.

**Comment by flohgh** [comment]
Also, the order given by the algorithm is actually doable (if not realistic) so for the seed you mentioned, the algorithm considers you can actually go to speed and wave locaction without hi/space jump.  Indeed you can do it from bubble mountain if you run from the save station room, jump very late, and wall jump as soon as you can. You can also go through the bottom as intended by the game, but it might be a lengthy hell run. 

It might be a good idea to add this bubble mountain jump as a technique.

**Comment by flohgh** [comment]
the latest release **greatly** improved solver pathing, closing the issue.
